driver-core/src/main/com/mongodb/internal/connection/ClientMetadataHelper.java

```diff
@@ -102,69 +72,88 @@ private static boolean nameMatches(final String name, final String... prefixes)
         return false;
     }
 
-    @Nullable
-    static BsonDocument createClientMetadataDocument(@Nullable final String applicationName) {
-        return createClientMetadataDocument(applicationName, null);
-    }
-
-    @Nullable
-    public static BsonDocument createClientMetadataDocument(@Nullable final String applicationName,
+    public static BsonDocument getClientMetadataDocument(@Nullable final String applicationName,
                                                             @Nullable final MongoDriverInformation mongoDriverInformation) {
-        return createClientMetadataDocument(applicationName, mongoDriverInformation, CLIENT_METADATA_DOCUMENT);
-    }
-
-    @Nullable
-    static BsonDocument createClientMetadataDocument(@Nullable final String applicationName,
-                                                     @Nullable final MongoDriverInformation mongoDriverInformation,
-                                                     final BsonDocument templateDocument) {
         if (applicationName != null) {
             isTrueArgument("applicationName UTF-8 encoding length <= 128",
                     applicationName.getBytes(StandardCharsets.UTF_8).length <= 128);
         }
 
-        BsonDocument document = templateDocument.clone();
-        if (applicationName != null) {
-            document.append(APPLICATION_FIELD, new BsonDocument(APPLICATION_NAME_FIELD, new BsonString(applicationName)));
-        }
+        // client fields are added in "preservation" order:
+        BsonDocument client = new BsonDocument();
+        tryWithLimit(client, d -> putAtPath(d, "application.name", applicationName));
+        MongoDriverInformation baseDriverInfor = getDriverInformation(null);
+        // required fields:
+        tryWithLimit(client, d -> {
+            putAtPath(d, "driver.name", listToString(baseDriverInfor.getDriverNames()));
+            putAtPath(d, "driver.version", listToString(baseDriverInfor.getDriverVersions()));
+        });
+        tryWithLimit(client, d -> putAtPath(d, "os.type", getOperatingSystemType(getOperatingSystemName())));
+        // full driver information:
+        MongoDriverInformation fullDriverInfo = getDriverInformation(mongoDriverInformation);
+        tryWithLimit(client, d -> {
+            putAtPath(d, "driver.name", listToString(fullDriverInfo.getDriverNames()));
+            putAtPath(d, "driver.version", listToString(fullDriverInfo.getDriverVersions()));
+        });
+        // optional fields:
```

Comment:
Suggest failing early if limit is reached. Perhaps adapt `tryWithLimit` so it could take multiple modifiers and fail fast.

eg: `tryWithLimit(final BsonDocument document, final Consumer<BsonDocument>... modifiers)` 

Then all these optional fields can be fail fast.

Reply:
I considered this, but if the preceding field is too large (let's say, 100 chars) but the ensuing fields are all small (say, 5 chars), we might omit the large field but still fit all the small fields.

Reply:
Interesting, is that per specification? Seems expensive to keep on trying to encode when you could fail each time.